### ISSUE TITLE
Syncing is still slow with ApplyOutOfSyncOnly=true

### ISSUE THREAD

**Issue by hintofbasil** [argoproj/argo-cd#5966]
If you are trying to resolve an environment-specific issue or have a one-off question about the edge case that does not require a feature then please consider asking a question in argocd slack [channel](https://argoproj.github.io/community/join-slack).

Checklist:

* [x] I've searched in the docs and FAQ for my answer: https://bit.ly/argocd-faq.
* [x] I've included steps to reproduce the bug.
* [x] I've pasted the output of `argocd version`.

**Describe the bug**

We have upgraded to ArgoCD v2.0.0 and set `ApplyOutOfSyncOnly=true` on our applications.  We are still seeing ArgoCD take a very long time to Sync when only a single resource is changed in each application.  We have also noticed that there are a surprising number of kubectl requests in the metrics when only ~30 resources should be applied, one to each cluster (see screenshots).

We noticed that during the sync the [dry run stage](https://github.com/argoproj/gitops-engine/blob/4342d853a34cd703c5a95ba9b4e62c8a907571b9/pkg/sync/sync_context.go#L378) does not filter down the resources before doing a dry run for each resource.  We added the `Validate=false` flag to each application to skip this stage and we are still seeing a very slow sync with a surprising number of kubectl requests.

**To Reproduce**

Create an application with a very large number of resources to deploy (we use ~1400).  Set the `ApplyOutOfSyncOnly=true` and `Validate=false` flags on the application.  Sync the application and monitor the sync time, and the number of kubectl requests made.

Our tests were performed when changing a single resource in ~40 applications each of size ~1400 resources.  Testing with a single application is likely to be faster - but still highlight the issue - as I suspect the bottle neck is the kubectlParallel limit which is shared across all applications.

**Expected behavior**

I expect the application to only make 1 kubectl request for each out of sync resource (2 if including the dry run).
I also expect the application to sync in less than 1 minute.

**Screenshots**

Screenshots taken while clusters were still syncing.  We set `--status-processors` to 50, `--operation-processors` to 25 and `--kubectl-parallelism-limit` to 30.

![image](https://user-images.githubusercontent.com/2506297/113563862-58dd5a00-9600-11eb-9c35-f8bbbc0ddbec.png)

![image](https://user-images.githubusercontent.com/2506297/113563812-43683000-9600-11eb-9143-b18c811553a4.png)

**Version**

```shell
argocd: v1.8.1+c2547dc.dirty
  BuildDate: 2020-12-10T04:44:20Z
  GitCommit: c2547dca95437fdbb4d1e984b0592e6b9110d37f
  GitTreeState: dirty
  GoVersion: go1.15.5
  Compiler: gc
  Platform: darwin/amd64
argocd-server: v2.0.0-rc2+9603ae3
  BuildDate: 
  GitCommit: 
  GitTreeState: 
  GoVersion: 
  Compiler: 
  Platform: 
  Ksonnet Version: 
  Kustomize Version: 
  Helm Version: 
  Kubectl Version: 
  Jsonnet Version: 

```

**Logs**

```
Paste any relevant application logs here.
```

**Comment by jessesuen** [comment]
> We noticed that during the sync the dry run stage does not filter down the resources before doing a dry run for each resource.

Your analysis is correct. We'll need to do additional filtering to avoid all of the dry-run apply. This may need to go in a v2.0.1 patch since we plan on releasing v2.0.0 on Wednesday.

**Comment by alexmt** [comment]
Fixed and cherry-picked into release-2.0 branch. Fix will be available in v2.0.0-rc4 (build is in progress: https://github.com/argoproj/argo-cd/runs/2273673184)

**Comment by hintofbasil** [comment]
This has reduced our deployment time down to 10 minutes for the same change tested above.  We are also seeing a significant reduction in the number of kubectl requests to the point it no longer looks like a bottleneck.

Thanks for pushing this through everyone!
